cub/cub/device/dispatch/dispatch_streaming_reduce.cuh

```diff
@@ -0,0 +1,451 @@
+  //! input items.
+  //!
+  //! @param[in] init
+  //!   The extremum value to be returned for empty problems
+  //!
+  //! @param[in] stream
+  //!   CUDA stream to launch kernels within.
+  //! @endrst
+  CUB_RUNTIME_FUNCTION _CCCL_FORCEINLINE static cudaError_t Dispatch(
+    void* d_temp_storage,
+    size_t& temp_storage_bytes,
+    InputIteratorT d_in,
+    OutputIteratorT d_result_out,
+    GlobalOffsetT num_items,
+    ReductionOpT reduce_op,
+    InitT init,
+    cudaStream_t stream)
+  {
+    // The input type
+    using input_value_t = cub::detail::value_t<InputIteratorT>;
+
+    // Constant iterator to provide the offset of the current partition for the user-provided input iterator
+    using constant_offset_it_t = cub::ConstantInputIterator<GlobalOffsetT>;
+
+    // Wrapped input iterator to produce index-value tuples, i.e., <PerPartitionOffsetT, InputT>-tuples
+    // We make sure to offset the user-provided input iterator by the current partition's offset
+    using arg_index_input_iterator_t =
+      ArgIndexInputIterator<offset_iterator<InputIteratorT, constant_offset_it_t>, PerPartitionOffsetT, InitT>;
+
+    // The type used for the aggregate that the user wants to find the extremum for
+    using output_aggregate_t = InitT;
+
+    // The output tuple type (i.e., extremum plus index tuples)
+    using per_partition_accum_t = KeyValuePair<PerPartitionOffsetT, output_aggregate_t>;
+    using global_accum_t        = KeyValuePair<GlobalOffsetT, output_aggregate_t>;
+
+    // Unary promotion operator type that is used to transform a per-partition result to a global result
+    // operator()(per_partition_accum_t) -> global_accum_t
+    using promote_to_global_op_t = promote_to_global_op<GlobalOffsetT>;
+
+    // Reduction operator type that enables accumulating per-partition results to a global reduction result
+    using accumulating_transform_output_op_t =
+      accumulating_transform_output_op<global_accum_t, promote_to_global_op_t, ReductionOpT, OutputIteratorT>;
+
+    // The output iterator that implements the logic to accumulate per-partition result to a global aggregate and,
+    // eventually, write to the user-provided output iterators
+    using accumulating_transform_out_it_t =
+      THRUST_NS_QUALIFIER::tabulate_output_iterator<accumulating_transform_output_op_t>;
+
+    // Empty problem initialization type
+    using empty_problem_init_t = empty_problem_init_t<per_partition_accum_t>;
+
+    // Per-partition DispatchReduce template specialization
+    using dispatch_reduce_t =
+      DispatchReduce<arg_index_input_iterator_t,
+                     accumulating_transform_out_it_t,
+                     PerPartitionOffsetT,
+                     ReductionOpT,
+                     empty_problem_init_t,
+                     per_partition_accum_t,
+                     PolicyChainT>;
+
+    void* allocations[2]       = {nullptr};
```

Comment:
Suggestion: even though the language rules would set the second array member to `nullptr` here, let's be explicit.
```suggestion
    void* allocations[2]       = {nullptr, nullptr};
```